client/components/community-translator/translatable.jsx

```diff
@@ -0,0 +1,214 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import { isEmpty } from 'lodash';
+import { translate } from 'i18n-calypso';
+import classNames from 'classnames';
+import Gridicon from 'gridicons';
+
+/**
+ * Internal dependencies
+ */
+import Dialog from 'components/dialog';
+import Button from 'components/button';
+import TranslatableField from './translatable-field';
+import TranslatedSuccess from './translated-success';
+import { getTranslationData, getTranslationPermaLink, submitTranslation } from './utils.js';
+
+export class Translatable extends Component {
+	state = {
+		showDialog: false,
+		originalData: {},
+		formState: {},
+	};
+
+	hasDataLoaded() {
+		return ! isEmpty( this.state.originalData );
+	}
+
+	handleTranslationChange = event => {
+		const { name, value } = event.target;
+		this.setState( {
+			formState: {
+				...this.state.formState,
+				[ name ]: value,
+			},
+			submitting: false,
+			submissionSuccess: false,
+		} );
+	};
+
+	closeDialog = () => {
+		this.setState( {
+			showDialog: false,
+			submissionSuccess: false,
+			formState: this.state.submissionSuccess ? {} : this.state.formState,
+			originalData: this.state.submissionSuccess ? {} : this.state.originalData,
+		} );
+	};
+
+	openDialog = event => {
+		event.preventDefault();
+
+		this.setState( { showDialog: true } );
+
+		const { singular, context, plural, locale } = this.props;
+		! this.hasDataLoaded() &&
+			getTranslationData( locale, { singular, context, plural } ).then( originalData =>
```

Comment:
We should also add a `catch()` expression to handle the error case here.

Reply:
A `catch()` exists, but [appended to the main request call](https://github.com/Automattic/wp-calypso/blob/727603e1e64975f544590fbbec50465671a1b9fa/client/components/community-translator/utils.js#L75).

Do you think it's better to handle it in the component itself?

I didn't think it was worth going full redux quite yet given the extra code.

Reply:
Thanks for pointing it out. I didn't notice it was there. Unfortunately, the way it is written is actually not correct.

From its resemblance to the `then()` expression:
```
.then( response => normalizeDetailsFromTranslationData( head( response.body ) ) )
.catch( error => normalizeDetailsFromTranslationData( error ) );
```
The original intention looks like to allow callers to chain the `catch` callback with a normalized data, just like we chain a `then` callback here. However, returning a value from the catch callback actually won't pass the value down to the next catch callback. e.g.
```
const foo = () => {
	return new Promise( ( resolve, reject ) => {
		setTimeout( () => {
			reject( 'hahaha' );
		}, 1000 );
	} );
};

foo()
	.catch( ( err ) => {
		console.log( 'the 1st err: ', err );
		return err;
	} )
	.catch( ( err ) => {
		console.log( 'the 2nd err: ', err );
	} );
```
Running this snippet through `node` or a browser console, it will only print `the 1st err: hahaha`. To make this chainable, it should be:
```
return request
		.post( glotPressUrl )
		.withCredentials()
		.send( postFormData )
		.then( response => normalizeDetailsFromTranslationData( head( response.body ) ) )
		.catch( error => { throw normalizeDetailsFromTranslationData( error ); } );
```

and then yes, I think we should handle it in the component, otherwise the normalized data will simply go into void :)

Reply:
You can see I didn't get around to testing everything. :)